### ISSUE TITLE
How to register custom nodes into a cluster created using Node Driver

### ISSUE THREAD

**Issue by lucernae** [rancher/rancher#30467]
**What kind of request is this (question/bug/enhancement/feature request):** Question

TL;DR:

I created a cluster using Rancher UI and Node Driver. I add new nodes using this node driver. However I also have some legacy nodes that I want to join into the cluster (manually, as if I create a cluster using RKE Custom Cluster and perform a rancher agent join command on the node). Since node driver uses RKE, I assume this is going to work. Are my assumptions right?

**Steps to reproduce (least amount of steps as possible):**

First of all, I just want to know if this is a good approach or if we have some other alternatives.

1. I created a cluster using Node Driver for hetzner made by community: https://github.com/JonasProgrammer/docker-machine-driver-hetzner/releases/download/2.1.0/docker-machine-driver-hetzner_2.1.0_linux_amd64.tar.gz

2. I can add several node pool and nodes using this node driver. Cluster created (I assume using RKE?). Everything works fine.

3. I have some legacy cloud nodes around. I want to join it to ease transition from legacy docker-compose deployment to a kubernetes-based manifest using the same node (so I can migrate the data easily). 

4. Since I saw provisioned nodes uses cattle node agent, I assume I can do the same with my legacy node.

5. I retrieve cattle token:

```bash
export SECRET_NAME=`kubectl -n cattle-system get secrets | grep cattle-credentials | awk '{print $1}'`
export CATTLE_TOKEN=`kubectl -n cattle-system get secret $SECRET_NAME -o jsonpath='{.data.token}' | base64 --decode`
echo $CATTLE_TOKEN
```

6. I retrieve cattle agent version

```bash
export CLUSTER_AGENT=`kubectl -n cattle-system get pods | grep cattle-cluster-agent | awk '{print $1}'`
export RANCHER_IMAGE=`kubectl -n cattle-system get pods $CLUSTER_AGENT -o jsonpath='{.spec.containers[0].image}'`
echo $RANCHER_IMAGE
```

7. I construct the join command and run it (`EXTRA_JOIN_ARGS` were used to define labels, and node name)

```bash
sudo docker run -d --privileged --restart=unless-stopped --net=host -v /etc/kubernetes:/etc/kubernetes -v /var/run:/var/run $RANCHER_IMAGE --server $RANCHER_SERVER_ADDRESS --token $CATTLE_TOKEN --worker $EXTRA_JOIN_ARGS
```

8. Node were recognized in Rancher UI with empty pool name. If I add more node, all these nodes were grouped inside this (legacy) pool with no name. K8s scheduling works. New nodes can accept pod requests, resolved services, and proxying nginx ingress.

**Result:**

The problem:

Kubelet and controlplane node (both the real host and kubelet container) can't access pods in the nodes joined using above join command. Let's say cluster CIDR: `10.42.0.0/16`. Node 1 were created using node driver with CIDR `10.42.1.0/24`. Node 2 were joined using join command (legacy node) with assigned CIDR `10.42.2.0/24`. 

Then:

### This works:

From node 2: `tracepath 10.42.0.1` was a success which is node 2 trying to access gateway of the master node.
From node 2: `tracepath 10.42.1.1` was a success which is node 2 trying to access gateway of the node 1 (worker node created by node driver).
From node 1: `tracepath 10.42.2.1` was a success which is node 1 trying to access gateway of the node 2 (worker node joined by join command).
From master node, or kubelet container inside the master node: `tracepath 10.42.1.1` was a success which is master node trying to access gateway of the node 1 (worker node created by node driver).

### This doesn't work:

From master node, or kubelet container inside the master node: `tracepath 10.42.2.1` was unsuccessful (no-reply) which is master node trying to access gateway of the node 2 (worker node joined by join command).

Because master node can't reach node 2, any attempt to access node 2 will fail. Like cert manager webhook if the webhook endpoint is in node 2, or kube-proxy if the service matches pods in node 2.

**Other details that may be helpful:**

I'm checking the `ip route` of each node to see if the route exists. The rule does exists, but for some reason tracepath failed.

On master node:
```
default via 172.31.1.1 dev eth0 proto dhcp src <REDACTED PUBLIC IP> metric 100 
10.42.0.4 dev cali30f787785d4 scope link 
10.42.0.5 dev cali4a4002d3b63 scope link 
10.42.1.0/24 via 10.42.1.0 dev flannel.1 onlink 
10.42.2.0/24 via 10.42.2.0 dev flannel.1 onlink 
172.17.0.0/16 dev docker0 proto kernel scope link src 172.17.0.1 linkdown 
172.31.1.1 dev eth0 proto dhcp scope link src  <REDACTED PUBLIC IP> metric 100
```

On node 1:
```
default via 172.31.1.1 dev eth0 proto dhcp src <REDACTED PUBLIC IP> metric 100 
10.42.0.0/24 via 10.42.0.0 dev flannel.1 onlink 
10.42.1.4 dev calif98b334d63b scope link  # calico assigned pod ip for this node, there are many of it
10.42.2.0/24 via 10.42.2.0 dev flannel.1 onlink 
172.17.0.0/16 dev docker0 proto kernel scope link src 172.17.0.1 linkdown 
172.31.1.1 dev eth0 proto dhcp scope link src <REDACTED PUBLIC IP> metric 100 
```

On node 2:
```
default via 172.31.1.1 dev eth0 
10.42.0.0/16 dev docker0  proto kernel  scope link  src 10.42.0.1 
10.42.1.0/24 via 10.42.1.0 dev flannel.1 onlink 
10.42.2.0/24 via 10.42.2.0 dev flannel.1 onlink 
10.42.2.2 dev cali5456c1400d9  scope link # calico assigned pod ip for this node, there are many of it
169.254.169.250 dev docker0 # docker bridge internal network
172.17.0.0/16 dev docker0  proto kernel  scope link  src 172.17.0.1 
172.31.1.1 dev eth0  scope link 
```

The only difference that I recognize is that in node 2, there's no mention of it's public ip. But why the only problematic one is just the master node?

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI): v2.4.7
- Installation option (single install/HA): HA using RKE with 3 identical nodes.

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported): Custom with Node Driver
- Machine type (cloud/VM/metal) and specifications (CPU/memory): Hetzner Cloud, minimum spec 2 vCPU, 4GB RAM, SSD
- Kubernetes version (use `kubectl version`): 

From Rancher kubectl web terminal
```
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.6", GitCommit:"d32e40e20d167e103faf894261614c5b45c44198", GitTreeState:"clean", BuildDate:"2020-05-20T13:16:24Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.3", GitCommit:"2e7996e3e2712684bc73f0dec0200d64eec7fe40", GitTreeState:"clean", BuildDate:"2020-05-20T12:43:34Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
```

- Docker version (use `docker version`):

Both master node and node 1:
```
Client: Docker Engine - Community
 Version:           19.03.11
 API version:       1.40
 Go version:        go1.13.10
 Git commit:        42e35e61f3
 Built:             Mon Jun  1 09:12:34 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.11
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.10
  Git commit:       42e35e61f3
  Built:            Mon Jun  1 09:11:07 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

Node 2 (from legacy cloud server)
```
Client:
 Version:      18.03.1-ce
 API version:  1.37
 Go version:   go1.9.5
 Git commit:   9ee9f40
 Built:        Thu Apr 26 07:17:20 2018
 OS/Arch:      linux/amd64
 Experimental: false
 Orchestrator: swarm

Server:
 Engine:
  Version:      18.03.1-ce
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.9.5
  Git commit:   9ee9f40
  Built:        Thu Apr 26 07:15:30 2018
  OS/Arch:      linux/amd64
  Experimental: false
```

**Comment by Jasstkn** [comment]
@lucernae Hi! I've faced with the same problem. Did you manage a solution for it?

**Comment by Jasstkn** [comment]
FYI, I found following solution: https://rancher.example.com/v3/clusters/CLUSTER_ID/clusterregistrationtokens, here you can find docker command

**Comment by lucernae** [comment]
Hi @Jasstkn, my original problem was not about adding the node (we can add it). But after adding the node, the networking of the pods in that node is not working perfectly. So, I was wondering if this procedure were officially supported or not.

> FYI, I found the following solution: https://rancher.example.com/v3/clusters/CLUSTER_ID/clusterregistrationtokens, here you can find docker command

This is much more neater than my procedure above. Thanks for letting me know.

**Comment by rayjasson98** [comment]
Hi @lucernae, were you eventually able to successfully join a custom node to the cluster?
